### PR TITLE
Schneider PowerTag support

### PR DESCRIPTION
**Pull request by filips**
This commit introduces minimal Schneider PowerTag support (https://github.com/Koenkk/zigbee2mqtt/issues/7975).

Known limitations:
 - Encryption is NOT supported for RX capable GPDs
 - Only total power + total energy implemented
 - Only tested with a Powertag Flex 3P+0

### CODE REVIEW

File: src/controller/controller.ts [hunk 1]
```diff
@@ -460,7 +460,7 @@ class Controller extends events.EventEmitter {
         // Green power devices dont' have a modelID, create a modelID based on the deviceID (=type)
         const modelID = `GreenPower_${payload.deviceID}`;
 
-        let device = Device.byIeeeAddr(ieeeAddr);
+        let device = Device.byIeeeAddr(ieeeAddr, true);
```
Comment (Koenkk) [inline src/controller/controller.ts:463]: Why include deleted here? (the added `true`)

File: src/controller/controller.ts [hunk 2]
```diff
@@ -473,6 +473,16 @@ class Controller extends events.EventEmitter {
             const deviceJoinedPayload: Events.DeviceJoinedPayload = {device};
             this.selfAndDeviceEmit(device, Events.Events.deviceJoined, deviceJoinedPayload);
 
+            const deviceInterviewPayload: Events.DeviceInterviewPayload = {status: 'successful', device};
+            this.selfAndDeviceEmit(device, Events.Events.deviceInterview, deviceInterviewPayload);
+        } else if (device.isDeleted) {
```
Comment (Koenkk) [inline src/controller/controller.ts:478]: Why was this logic added? How can the device become deleted?

**Comment by Koenkk** [comment]
I will check if this doesn't break anything with existing green power device like enocean/hue tap, could you fix the test coverage in the mean time? You can run it locally by executing `npm run test-with-coverage`, results can be found in `coverage/lcov-report/index.html`

**Comment by filips** [comment]
Sure, sounds good. I'll get the test coverage fixed.

File: src/controller/controller.ts (hunk 2 above)
Comment (filips) [reply src/controller/controller.ts:478]: When the device is removed from the UI, and then re-paired without restarting Z2M.

**Comment by filips** [comment]
Test coverage fixed. I have in a couple of places used istanbul ignore - please let me know if you think any of those are not ok to skip.

File: src/controller/model/device.ts [hunk 3]
```diff
@@ -712,6 +714,10 @@ class Device extends Entity {
         }
     }
 
+    public async getNetworkParameters(): Promise<NetworkParameters> {
+        return Entity.adapter.getNetworkParameters();
```
Comment (Koenkk) [inline src/controller/model/device.ts:718]: This only seems to be used in a test case: https://github.com/Koenkk/zigbee-herdsman/pull/518/files#diff-143091d444c5f0ff7479e674fcbc691161665ccf435ad8c8c7d8f55d20b8361aR3924

Would be nice if this method can be removed

**Comment by Koenkk** [comment]
Tested & works, thanks!

**Comment by Inrego** [comment]
Looking forward to trying it! Unfortunately, I can't update until https://github.com/Koenkk/zigbee2mqtt/issues/12031 has been closed.

File: src/controller/model/device.ts (hunk 3 above)
Comment (filips) [reply src/controller/model/device.ts]: @Koenkk Hmm slight blunder from my side - I use the getNetworkParameters() from the converter: https://github.com/Koenkk/zigbee-herdsman-converters/blob/09495aca82b6498d2222664b7edff5cf78b70a7d/devices/schneider_electric.js#L57

Unless there is a better way of getting the network channel  from a converter I think we should re-add the method.

**Comment by smulle48** [comment]
Working great 👍 

Don't know if you already knew it, but I can see it's reporting the active power for each phase :-)

I've hooked it up to my PEV charger.

Before charging:

`Apr 13 21:01:29 homeassistant ea04f43b2a5b[535]: Zigbee2MQTT:debug 2022-04-13 21:01:29: Received Zigbee message from 'Clever lader', type 'commandNotification', cluster 'greenPower', data '{"commandFrame":{"attributes":{"acFrequency":500,"acFrequencyDivisor":10,**"activePower":0,"activePowerPhB":0,"activePowerPhC":0**,"powerDivisor":1000,"totalActivePower":0,"totalApparentPower":0},"clusterID":2820,"manufacturerCode":4190},"commandID":161,"frameCounter":84,"gppGddLink":255,"gppNwkAddr":4662,"options":20496,"payloadSize":50,"srcID":3791834163}' from endpoint 242 with groupID 52275`

Activated charger:

`Apr 13 21:07:58 homeassistant ea04f43b2a5b[535]: Zigbee2MQTT:debug 2022-04-13 21:07:58: Received Zigbee message from 'Clever lader', type 'commandNotification', cluster 'greenPower', data '{"commandFrame":{"attributes":{"acFrequency":500,"acFrequencyDivisor":10,**"activePower":3348,"activePowerPhB":3360,"activePowerPhC":3324**,"powerDivisor":1000,"totalActivePower":10032,"totalApparentPower":10611},"clusterID":2820,"manufacturerCode":4190},"commandID":161,"frameCounter":64,"gppGddLink":255,"gppNwkAddr":34557,"options":20496,"payloadSize":50,"srcID":3791834163}' from endpoint 242 with groupID 52275`

So at least I can see it in the debug log :-D

**Comment by filips** [comment]
> Working great 👍
> 
> Don't know if you already knew it, but I can see it's reporting the active power for each phase :-)
> 
> I've hooked it up to my PEV charger.
> 
> Before charging:
> 
> ```
> Apr 13 21:01:29 homeassistant ea04f43b2a5b[535]: Zigbee2MQTT:debug 2022-04-13 21:01:29: Received Zigbee message from 'Clever lader', type 'commandNotification', cluster 'greenPower', data '{"commandFrame":{"attributes":{"acFrequency":500,"acFrequencyDivisor":10,"activePower":0,"activePowerPhB":0,"activePowerPhC":0,"powerDivisor":1000,"totalActivePower":0,"totalApparentPower":0},"clusterID":2820,"manufacturerCode":4190},"commandID":161,"frameCounter":84,"gppGddLink":255,"gppNwkAddr":4662,"options":20496,"payloadSize":50,"srcID":3791834163}' from endpoint 242 with groupID 52275
> ```
> 
> Activated charger:
> 
> ```
> Apr 13 21:07:58 homeassistant ea04f43b2a5b[535]: Zigbee2MQTT:debug 2022-04-13 21:07:58: Received Zigbee message from 'Clever lader', type 'commandNotification', cluster 'greenPower', data '{"commandFrame":{"attributes":{"acFrequency":500,"acFrequencyDivisor":10,"activePower":3348,"activePowerPhB":3360,"activePowerPhC":3324,"powerDivisor":1000,"totalActivePower":10032,"totalApparentPower":10611},"clusterID":2820,"manufacturerCode":4190},"commandID":161,"frameCounter":64,"gppGddLink":255,"gppNwkAddr":34557,"options":20496,"payloadSize":50,"srcID":3791834163}' from endpoint 242 with groupID 52275
> ```
> 
> So at least I can see it in the debug log :-D

Good to hear! Yeah it is being read out - all the additional fields such as individual phase power, voltage, power factor etc. just need to be added in the converter.

**Comment by smulle48** [comment]
> > Working great 👍
> > Don't know if you already knew it, but I can see it's reporting the active power for each phase :-)
> > I've hooked it up to my PEV charger.
> > Before charging:
> > ```
> > Apr 13 21:01:29 homeassistant ea04f43b2a5b[535]: Zigbee2MQTT:debug 2022-04-13 21:01:29: Received Zigbee message from 'Clever lader', type 'commandNotification', cluster 'greenPower', data '{"commandFrame":{"attributes":{"acFrequency":500,"acFrequencyDivisor":10,"activePower":0,"activePowerPhB":0,"activePowerPhC":0,"powerDivisor":1000,"totalActivePower":0,"totalApparentPower":0},"clusterID":2820,"manufacturerCode":4190},"commandID":161,"frameCounter":84,"gppGddLink":255,"gppNwkAddr":4662,"options":20496,"payloadSize":50,"srcID":3791834163}' from endpoint 242 with groupID 52275
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Activated charger:
> > ```
> > Apr 13 21:07:58 homeassistant ea04f43b2a5b[535]: Zigbee2MQTT:debug 2022-04-13 21:07:58: Received Zigbee message from 'Clever lader', type 'commandNotification', cluster 'greenPower', data '{"commandFrame":{"attributes":{"acFrequency":500,"acFrequencyDivisor":10,"activePower":3348,"activePowerPhB":3360,"activePowerPhC":3324,"powerDivisor":1000,"totalActivePower":10032,"totalApparentPower":10611},"clusterID":2820,"manufacturerCode":4190},"commandID":161,"frameCounter":64,"gppGddLink":255,"gppNwkAddr":34557,"options":20496,"payloadSize":50,"srcID":3791834163}' from endpoint 242 with groupID 52275
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So at least I can see it in the debug log :-D
> 
> Good to hear! Yeah it is being read out - all the additional fields such as individual phase power, voltage, power factor etc. just need to be added in the converter.

That's a bit above my skills :-D

**Comment by filips** [comment]
> > > Working great +1
> > > Don't know if you already knew it, but I can see it's reporting the active power for each phase :-)
> > > I've hooked it up to my PEV charger.
> > > Before charging:
> > > ```
> > > Apr 13 21:01:29 homeassistant ea04f43b2a5b[535]: Zigbee2MQTT:debug 2022-04-13 21:01:29: Received Zigbee message from 'Clever lader', type 'commandNotification', cluster 'greenPower', data '{"commandFrame":{"attributes":{"acFrequency":500,"acFrequencyDivisor":10,"activePower":0,"activePowerPhB":0,"activePowerPhC":0,"powerDivisor":1000,"totalActivePower":0,"totalApparentPower":0},"clusterID":2820,"manufacturerCode":4190},"commandID":161,"frameCounter":84,"gppGddLink":255,"gppNwkAddr":4662,"options":20496,"payloadSize":50,"srcID":3791834163}' from endpoint 242 with groupID 52275
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Activated charger:
> > > ```
> > > Apr 13 21:07:58 homeassistant ea04f43b2a5b[535]: Zigbee2MQTT:debug 2022-04-13 21:07:58: Received Zigbee message from 'Clever lader', type 'commandNotification', cluster 'greenPower', data '{"commandFrame":{"attributes":{"acFrequency":500,"acFrequencyDivisor":10,"activePower":3348,"activePowerPhB":3360,"activePowerPhC":3324,"powerDivisor":1000,"totalActivePower":10032,"totalApparentPower":10611},"clusterID":2820,"manufacturerCode":4190},"commandID":161,"frameCounter":64,"gppGddLink":255,"gppNwkAddr":34557,"options":20496,"payloadSize":50,"srcID":3791834163}' from endpoint 242 with groupID 52275
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > So at least I can see it in the debug log :-D
> > 
> > 
> > Good to hear! Yeah it is being read out - all the additional fields such as individual phase power, voltage, power factor etc. just need to be added in the converter.
> 
> That's a bit above my skills :-D

I actually did it earlier today (https://github.com/Koenkk/zigbee-herdsman-converters/pull/4134), should be available on the dev branch now.

**Comment by smulle48** [comment]
> > > > Working great +1
> > > > Don't know if you already knew it, but I can see it's reporting the active power for each phase :-)
> > > > I've hooked it up to my PEV charger.
> > > > Before charging:
> > > > ```
> > > > Apr 13 21:01:29 homeassistant ea04f43b2a5b[535]: Zigbee2MQTT:debug 2022-04-13 21:01:29: Received Zigbee message from 'Clever lader', type 'commandNotification', cluster 'greenPower', data '{"commandFrame":{"attributes":{"acFrequency":500,"acFrequencyDivisor":10,"activePower":0,"activePowerPhB":0,"activePowerPhC":0,"powerDivisor":1000,"totalActivePower":0,"totalApparentPower":0},"clusterID":2820,"manufacturerCode":4190},"commandID":161,"frameCounter":84,"gppGddLink":255,"gppNwkAddr":4662,"options":20496,"payloadSize":50,"srcID":3791834163}' from endpoint 242 with groupID 52275
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > Activated charger:
> > > > ```
> > > > Apr 13 21:07:58 homeassistant ea04f43b2a5b[535]: Zigbee2MQTT:debug 2022-04-13 21:07:58: Received Zigbee message from 'Clever lader', type 'commandNotification', cluster 'greenPower', data '{"commandFrame":{"attributes":{"acFrequency":500,"acFrequencyDivisor":10,"activePower":3348,"activePowerPhB":3360,"activePowerPhC":3324,"powerDivisor":1000,"totalActivePower":10032,"totalApparentPower":10611},"clusterID":2820,"manufacturerCode":4190},"commandID":161,"frameCounter":64,"gppGddLink":255,"gppNwkAddr":34557,"options":20496,"payloadSize":50,"srcID":3791834163}' from endpoint 242 with groupID 52275
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > So at least I can see it in the debug log :-D
> > > 
> > > 
> > > Good to hear! Yeah it is being read out - all the additional fields such as individual phase power, voltage, power factor etc. just need to be added in the converter.
> > 
> > 
> > That's a bit above my skills :-D
> 
> I actually did it earlier today ([Koenkk/zigbee-herdsman-converters#4134](https://github.com/Koenkk/zigbee-herdsman-converters/pull/4134)), should be available on the dev branch now.

Wow....sweet :-)

Thank you very much. :-D

I was on the right track, I can see, but it's a pain in the a**, as I'm using the z2m home assistant plugin.

Have to move it to my test setup every time I want to test something :-/

**Comment by Inrego** [comment]
Not quite sure of the reason, but my Z2M keeps crashing when pairing one of these devices. It does pair to Z2M, but does not set up any reporting (so no values). And my zigbee network just mostly doesn't work, and Z2M crashes very frequently.

**Comment by smulle48** [comment]
> Not quite sure of the reason, but my Z2M keeps crashing when pairing one of these devices. It does pair to Z2M, but does not set up any reporting (so no values). And my zigbee network just mostly doesn't work, and Z2M crashes very frequently.

How does your debug log look like when adding the device?

**Comment by filips** [comment]
What type of zigbee coordinator do you use? Currently investigating some issues with the Conbee II and green power..

**Comment by Inrego** [comment]
Conbee II :D

**Comment by Inrego** [comment]
> What type of zigbee coordinator do you use? Currently investigating some issues with the Conbee II and green power..

Which coordinator have you had success with?

**Comment by smulle48** [comment]
> > What type of zigbee coordinator do you use? Currently investigating some issues with the Conbee II and green power..
> 
> Which coordinator have you had success with?

I've been using conbee II myself before. Even with deCONZ it was rubbish 😀 

Now I'm using the slae.sh adapter and it's working great 👍 

https://slae.sh/projects/cc2652/

**Comment by filips** [comment]
I'm also using the slae.sh adapter which is working great for me. I would expect any CC2652 based adapter to work similarly.

**Comment by Inrego** [comment]
So the Sonoff 3.0 dongle for example? Seems half the price, plus it's sold in my country so faster/cheaper shipping

**Comment by filips** [comment]
> So the Sonoff 3.0 dongle for example? Seems half the price, plus it's sold in my country so faster/cheaper shipping

Yeah it should work, although i have not tried it.

**Comment by smulle48** [comment]
Hi all

First of all, this is working very good.

Today when I looked at it while charging my car, I noticed something. Could be the reason that I did not pay enough attention in school, but something caught my eyes :-D

On "full speed" I measure around 10.000w total, with around 3300w on each phase and around 230v on all phases.

But then when I look at the amps, it says 15a on each phase. Is that correct? is the amp not calculated from each phase watt?

The correct amp pr phase would be around 5.5a right?

clever 1
clever 2

**Pull request merged**